### PR TITLE
XS✔ ◾ 🐛 Bug - Exclude deleted Azure DevOps PBIs from duplicate detection (#862)

### PR DESCRIPTION
**Pull request by calumjs**
Fixes #862

## Root cause
When YakShaver records a YakShave, the agent searches the backlog for an existing/duplicate work item before creating a new one. The prompt that drives this agent (`SHARED_ISSUE_CREATION_RULES`, injected into both the default project prompt and the default custom prompt) gave **no guidance about deleted items**. So a previously created PBI that had since been **deleted** in Azure DevOps (`System.State = "Removed"`) was still returned by the search and treated as a live duplicate.

The agent then tried to **update** that deleted PBI — which Azure DevOps rejects — and fell back to creating a new, **incomplete** item carrying only a title plus a "duplicate" comment, dropping the steps-to-reproduce and acceptance criteria.

## Fix
Added an explicit `DUPLICATE_DETECTION_RULES` section to the shared issue-creation rules in `src/backend/constants/prompts.ts`. It instructs the agent to:
- **Ignore deleted/removed items** during duplicate detection — a deleted item does NOT count as a duplicate (AC #1). Covers Azure DevOps "Removed"/recycle-bin items, GitHub deleted / closed-as-not-planned issues, deleted Jira issues, and anything a tool reports as deleted/removed/archived/not found.
- When querying Azure DevOps with WIQL, **exclude removed items** (`AND [System.State] <> 'Removed'`) (AC #1).
- **Never attempt to update** a deleted/removed item (AC #2).
- When the only matching item is deleted, create a **brand-new, fully-populated** item — title, steps to reproduce, acceptance criteria (AC #3, #4) — and **do NOT add a "duplicate" comment** (AC #5).

Because the rule is composed into both `defaultProjectPrompt` (remote/portal flow) and `defaultCustomPrompt` (local flow), every issue-creation path now carries the guidance.

## Test
Added `src/backend/constants/prompts.test.ts` — deterministic, dependency-free string assertions, one per acceptance criterion, plus wiring checks that the guidance is present in both prompt paths. 8 tests, all passing.

## Validation
- `npm run build` (tsc): green
- `npx vitest run src/backend/constants/prompts.test.ts`: 8/8 passing
- `npx @biomejs/biome@2.3.11 check`: clean on changed files

## Live repro / walkthrough — PENDING ⚠️
A video fork currently owns the running app, so live reproduction and the before/after walkthrough were deferred. Still needs live verification:
1. Create + record a YakShave that creates a PBI in Azure DevOps; delete that PBI; record another YakShave for the same issue and confirm a **fully populated** new PBI is created with no duplicate comment and no attempt to update the deleted item.

Note: this is a prompt/guidance fix that steers the LLM agent. The deterministic tests lock in the instruction text; the end-to-end behaviour should be confirmed against a live Azure DevOps board.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by calumjs — COMMENTED** [review]
## Review: duplicate-detection guidance for deleted Azure DevOps PBIs (#862)

The fix adds `DUPLICATE_DETECTION_RULES` to `SHARED_ISSUE_CREATION_RULES` and wires it into the two default-prompt constants. The intent is sound, but as it stands the guarantee the PR claims is not actually established. There are two distinct problems: the change is not verified end-to-end, and it does not reach the prompts most real projects actually run.

### High

**1. The fix is only verified by string assertions; behaviour is never exercised.**
`src/backend/constants/prompts.test.ts` is entirely substring checks against the prompt *constants* (`toContain("deleted")`, `toContain("[System.State] <> 'Removed'")`, `toMatch(/never attempt to update.../i)`, plus a wiring check that `DUPLICATE_DETECTION_RULES` is embedded). These prove the guidance text was written and wired in — not that the agent behaves differently. The entire fix is a behavioural change to an LLM agent: there is no code-level duplicate-detection logic in the backend (no `System.State` / WIQL / `isDeleted` dedup path in `shave-service.ts` or elsewhere), so detection and the update-vs-create decision live entirely in the model. String tests cannot stand in for that. The PR body itself marks live verification as PENDING.

Before merge, run the live repro the PR body already specifies, and capture both outcomes:
- **Bug path:** record a YakShave that creates a real Azure DevOps PBI, delete it, record again — confirm a fully populated new PBI with no "duplicate" comment and no update attempt on the removed item.
- **Happy-path twin (currently missing):** record a YakShave whose match is a LIVE, non-removed duplicate and confirm the agent STILL detects and updates it. This proves the new rules did not over-trigger and suppress legitimate dedup. Without this twin, the change is verified on only one branch.

**2. The guidance only reaches default prompts; existing project/custom/remote prompts bypass it.**
In `src/backend/services/workflow/prompt-selection-service.ts:52`, the active prompt is resolved as `projectDetails.desktopAgentProjectPrompt ||= defaultProjectPrompt`. The default (the only carrier of the new rules) is substituted ONLY when the resolved prompt is empty. Any project with a non-empty prompt short-circuits it:
- Local user-authored prompts map `localPrompt.content` straight through (`prompt-manager.ts:186`).
- Remote/portal projects return the portal payload's `desktopAgentProjectPrompt` verbatim (`prompt-manager.ts:251`).

The base orchestrator system prompt (`src/backend/services/mcp/prompts.ts`) carries no issue-creation guidance, so the rules reach the agent only if they are inside the resolved project prompt (`mcp-orchestrator.ts:252-254`). Decisively, issue #862 was reported against a real SSW *portal* board — exactly the remote/portal class that bypasses the default when it supplies its own prompt. So the PR body's claim that "every issue-creation path now carries the guidance" is overstated, and the original bug can still recur for these users.

Remedy: append `SHARED_ISSUE_CREATION_RULES` (or at least `DUPLICATE_DETECTION_RULES`) at composition time onto the finally-resolved prompt in `prompt-selection-service`, rather than only into the static defaults — or normalize/migrate loaded prompts so every issue-creation prompt carries the rules regardless of source. Add a test that selects a non-empty custom `desktopAgentProjectPrompt` and asserts the final orchestrator prompt still contains the duplicate rules (the `||=` branch is currently untested).

### Medium

**3. Existing stored custom prompts are frozen snapshots and never re-composed.**
`src/backend/services/storage/custom-prompt-storage.ts:26` builds the template content from `${defaultCustomPrompt}` at module-load time, so a NEWLY created template-derived prompt carries the guidance. But saved prompts in `custom-settings.enc` are returned verbatim by `getAllPrompts()` with no re-composition. Users who already have a saved custom prompt (or any free-form prompt not derived from the template) will not receive the deleted-item guidance and remain exposed to #862 — the precise population the fix is meant to protect. This is covered by the same runtime-append remedy as finding 2; otherwise migrate/re-compose stored prompts, or explicitly document the limitation. No test round-trips a stored-then-retrieved prompt, so this gap is untested.

### Summary
The prompt text and its wiring into the defaults are correct, but (a) the behaviour is unproven end-to-end and the happy-path/dedup-still-works branch is never exercised, and (b) the runtime resolution applies the guidance only to the fallback default, leaving local-custom and remote/portal prompts — including the reporter's own portal board — uncovered. Recommend addressing both before merge, or scoping the PR claim to "default/template prompts only" and following up.

File: src/backend/constants/prompts.ts [hunk 1]
```diff
@@ -1,3 +1,22 @@
+/**
+ * Duplicate-detection guidance shared by every issue-creation prompt.
```
Comment (calumjs) [inline src/backend/constants/prompts.ts:2]: Finding (high): this new `DUPLICATE_DETECTION_RULES` block is guidance to the LLM agent, and the only tests added (prompts.test.ts) assert its substrings are present in the default-prompt constants. That proves the text was wired in, not that the agent obeys it. There is no code-level dedup logic to unit-test, so the bug path AND the happy-path twin (a LIVE non-removed duplicate is STILL detected + updated) need a live Azure DevOps repro before merge. Also note these rules only reach prompts that fall back to the default (prompt-selection-service.ts:52 `||=`); local-custom and remote/portal prompts bypass them. See the summary review for the full write-up.

Comment (calumjs) [reply src/backend/constants/prompts.ts:2]: Addressed the coverage-gap half of this finding in fd4e5e8.

**Root cause confirmed and fixed.** The rules were only baked into the two static default-prompt constants, and the runtime resolution (`prompt-selection-service.ts:52` `desktopAgentProjectPrompt ||= defaultProjectPrompt`) only substitutes a default when the selected project has *no* stored prompt. So any local-custom or remote/portal project shipping its own `desktopAgentProjectPrompt` — including prompts saved before this fix, which is the path the original report (#862) came in on (an SSW portal Azure DevOps board) — bypassed the guidance entirely.

**Fix:** added `ensureDuplicateDetectionRules(prompt)` in `constants/prompts.ts` and applied it to the finally-resolved `desktopAgentProjectPrompt` in `prompt-selection-service.ts`, right after the `||=` fallback. It appends `DUPLICATE_DETECTION_RULES` when absent and is a no-op when already present (defaults / template-derived prompts), so the guidance now reaches **all three sources** — default, stored local-custom, and remote portal — regardless of when the prompt was created. This is the single composition-time choke point every issue-creation path funnels through, so I did not need to touch the orchestrators or the checkpoint-resume path (checkpoints are written from this same resolved value).

**Added tests** for the previously-untested resolution branch: a non-empty custom prompt gets the rules appended, idempotency, no double-append on repeated application, and empty/undefined pass-through.

**On the live-repro half:** that remains a human-gated step — it needs the running Electron app against a live Azure DevOps board and can't be driven deterministically in CI. The behavioural change is still LLM guidance, so the string/branch tests prove the rules are *delivered to the agent on every path* (which the fix above now guarantees), not that the model obeys them. The bug path and the happy-path twin (a LIVE non-removed duplicate is still detected + updated) should be confirmed live before merge as the PR body notes.

Validation on this change: `npm run build` ✅, `npx vitest run` non-DB suite ✅ (176 passed; the only failures are pre-existing native-binary issues on this win32-arm64 box — better-sqlite3 DB tests and ffmpeg-static), `npx biome check` ✅ on all three changed files (the one repo-wide biome error is a pre-existing import-ordering issue in `VideoPreviewModal.tsx`, untouched by this PR).

**Review by calumjs — COMMENTED** [review]
## Review — verification gap on a prompt-only behavioural fix

This PR fixes #862 (a deleted Azure DevOps PBI was still treated as a live duplicate) entirely by appending natural-language guidance (`DUPLICATE_DETECTION_RULES`) to the issue-creation system prompt and composing it in via `ensureDuplicateDetectionRules()`. There is no deterministic code path: grepping the backend shows no `System.State`/`Removed`/`isDeleted`/WIQL logic outside the prompt strings — all work-item search/create/update is driven by the LLM agent through MCP tools. The behavioural change therefore lives entirely in the model obeying the new prose.

The `prompts.test.ts` suite (12 tests) is exclusively string assertions over the prompt constant (`toContain("deleted")`, `toContain("[System.State] <> 'Removed'")`, `toMatch(/never attempt to update.../i)`) plus wiring and `ensureDuplicateDetectionRules` idempotency checks. These prove the instruction text was written and wired in — a fine lock-in — but cannot prove the agent's behaviour changed. The PR body itself marks **Live repro / walkthrough — PENDING**, so the end-to-end behaviour is unverified.

### High

- **One-sided verification — no behavioural proof on either branch.** Every acceptance criterion is behavioural (deleted item excluded from dedup, no update of the removed item, fully-populated new PBI, no "duplicate" comment) yet is validated only by asserting the prompt text exists. Critically, only the BUG branch is reasoned about; the **happy-path twin is never exercised** — a LIVE, non-removed duplicate must STILL be detected and updated. The new rule broadens "not a duplicate" to "anything a tool reports as deleted, removed, archived, **or not found**", which an LLM could plausibly over-apply and start suppressing legitimate dedup. Nothing proves the rule does not over-trigger.
  - **Before merge, run BOTH branches end-to-end and capture them on the PR:**
    - *(bug)* Create a real ADO PBI via YakShave -> delete it (`System.State = Removed`) -> re-shave the same issue -> confirm a brand-new fully-populated PBI with no update attempt against the removed item and no "duplicate" comment.
    - *(twin)* Re-shave against a LIVE, non-removed duplicate -> confirm the agent STILL detects and updates it, proving no over-trigger.

### Medium

- **Static trace, not a live run — AC #1 (WIQL `[System.State] <> 'Removed'`) relies solely on LLM compliance.** There is no enforced code path that filters Removed results, blocks an update against a removed work item, or forces full population; the WIQL is composed by the model. Prompt-only fixes regress silently when the model changes.
  - Primary remedy: the live ADO E2E above, capturing that the actual tool call excludes Removed items.
  - Optional defensive measure: enforce the Removed filter / refuse-update-on-deleted in the ADO MCP/tool layer rather than relying entirely on prompt adherence. (Noted as harder here since the desktop drives ADO only via MCP tools — it would require intercepting tool I/O — hence "consider", not "required".)

Net: the diff is a reasonable instruction lock-in, but the behavioural ACs are unverified. Please complete the live repro the PR body already specifies, plus the missing live-duplicate twin, before merge.

File: src/backend/constants/prompts.ts [hunk 2]
```diff
@@ -1,3 +1,40 @@
...
+ * The agent then tried to UPDATE the deleted item — which the platform rejects — and fell back to
+ * creating a new, incomplete item carrying only a title plus a "duplicate" comment, dropping the
+ * steps-to-reproduce and acceptance criteria. Deleted items must NEVER count as a match: exclude
+ * them from the search, never update them, and never add a duplicate comment when the only match
+ * is deleted.
+ */
+export const DUPLICATE_DETECTION_RULES = `10) **Duplicate Detection (CRITICAL)**:
+- Before creating an item you may check the backlog for an existing duplicate, but you MUST IGNORE items that have been deleted or removed.
+- A deleted/removed item DOES NOT count as a duplicate. Treat the following as deleted and exclude them: Azure DevOps work items in the "Removed" state or returned as deleted/in the recycle bin (\`System.State\` = "Removed", or \`isDeleted\` true); GitHub issues that are deleted or closed as "not planned"; Jira issues that are deleted; and any item a tool reports as deleted, removed, archived, or not found.
+- When querying Azure DevOps with WIQL, exclude removed items, e.g. add \`AND [System.State] <> 'Removed'\` to the query.
```
Comment (calumjs) [inline src/backend/constants/prompts.ts]: AC #1 is enforced only by this prose — there is no code path that filters `System.State = Removed` from the dedup query; the WIQL is composed by the LLM agent. Confirm against a live ADO board that the actual tool call excludes Removed items before merge (and consider enforcing the filter in the MCP/tool layer so the fix does not regress silently when the model changes).

File: src/backend/constants/prompts.test.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    expect(rules).toContain("removed");
+    expect(rules).toMatch(/does not count as a duplicate/i);
+  });
+
+  it("names the Azure DevOps 'Removed' state so removed work items are excluded (AC #1)", () => {
+    // AC #1: deleted PBIs are excluded from duplicate detection queries against Azure DevOps.
+    expect(DUPLICATE_DETECTION_RULES).toContain("Removed");
+    expect(DUPLICATE_DETECTION_RULES).toContain("System.State");
+    // A concrete WIQL filter the agent can apply.
+    expect(DUPLICATE_DETECTION_RULES).toContain("[System.State] <> 'Removed'");
```
Comment (calumjs) [inline src/backend/constants/prompts.test.ts:29]: This (and the other 11 tests here) is a string assertion over the prompt constant — it proves the WIQL guidance text exists, not that the agent emits the filter or that behaviour changed. The behavioural ACs (deleted item excluded, no update of the removed item, fully-populated new PBI, no duplicate comment) are unverified, and the happy-path twin (a LIVE non-removed duplicate is STILL detected + updated, i.e. the rule does not over-trigger) is never exercised. Run both branches end-to-end against a live ADO board and capture them on the PR before merge.

File: src/backend/constants/prompts.ts (hunk 1 above)
Comment (calumjs) [reply src/backend/constants/prompts.ts:2]: Addressed the happy-path-twin / over-trigger half of this finding in 93782b9.

The prompt was one-sided: it told the agent to IGNORE deleted/removed items but never stated the symmetric rule, so nothing structurally prevented the LLM from over-applying "exclude removed" and suppressing legitimate dedup of LIVE items. `DUPLICATE_DETECTION_RULES` now spells out both sides explicitly: only deleted/removed items are excluded; any item in any other (non-removed) state that matches is STILL a duplicate and must be updated, never re-created; and the WIQL `Removed` filter must drop ONLY removed items, not live items in other states. Added happy-path-twin tests locking that in (live duplicate still detected/updated; WIQL exclusion scoped to removed only) — 14/14 in prompts.test.ts.

On the live-ADO end-to-end (bug path + the live-duplicate twin run against a real board): that remains the human-gated step the PR body flags. It needs the running Electron app against a live Azure DevOps board and can't be driven deterministically here (the running app is held by another fork, same constraint the review hit). The deterministic change now makes the steering two-sided rather than relying on the model to infer the symmetric case.

I deliberately did NOT add a structural tool-layer guard (intercept/parse work-item tool I/O to block updates on Removed items): work-item search/update is mediated entirely by dynamic per-provider MCP tools, so that guard would be a brittle cross-provider heuristic that could also break legitimate updates — riskier than the prompt approach, which is why the review itself flagged it as optional.

Validation: `npm run build` ✅, `npx vitest run` non-DB ✅ (178 passed), `npx biome check` ✅ on both changed files. (The better-sqlite3 DB and ffmpeg suites fail to load on this win32-arm64 box — pre-existing native-binary issue, unrelated to this change.)

File: src/backend/constants/prompts.ts (hunk 2 above)
Comment (calumjs) [reply src/backend/constants/prompts.ts]: Addressed the deterministic half in 93782b9 and skipping the tool-layer-enforcement suggestion with reason.

I strengthened the WIQL guidance so it is two-sided: the `AND [System.State] <> 'Removed'` filter must exclude ONLY removed items and must not drop live items in other states, and a live match is still a duplicate to be updated. Added tests asserting the WIQL exclusion is scoped to removed-only. That removes the over-trigger ambiguity but, as you note, AC #1 is still ultimately emitted by the agent.

On enforcing the `Removed` filter in the MCP/tool layer: skipped. The desktop never calls the ADO work-item API directly — search/create/update all flow through dynamic, per-provider MCP tools (`ado__*`, `github__*`, Jira, etc.) whose result schemas differ. A guard that intercepts tool I/O to detect/inject a Removed filter or block updates on removed items would be exactly the kind of brittle, provider-specific heuristic the repo's robustness skill warns against, and it could break legitimate updates. The two-sided prompt is the more robust lever here.

The live-ADO confirmation (actual tool call excludes Removed items) remains the human-gated step the PR body flags — not runnable deterministically in this environment.

Validation: build ✅, non-DB vitest ✅ (178), biome ✅.

File: src/backend/constants/prompts.test.ts (hunk 3 above)
Comment (calumjs) [reply src/backend/constants/prompts.test.ts:29]: Addressed the missing happy-path twin (deterministically) in 93782b9.

You're right that the string assertions prove text-presence, not agent behaviour. The most actionable gap you named — the happy-path twin where a LIVE non-removed duplicate must STILL be detected and updated (i.e. the rule does not over-trigger) — was not even present in the prompt to assert. I made `DUPLICATE_DETECTION_RULES` two-sided (only deleted/removed excluded; live matches still updated, never re-created; WIQL exclusion scoped to removed-only) and added the twin tests here: `still requires a LIVE (non-removed) duplicate to be detected and updated` and `scopes the WIQL exclusion to removed items only`. prompts.test.ts is now 14/14.

These remain string/lock-in tests by nature — there is no code-level dedup path to unit-test — so the four behavioural ACs plus the live-duplicate twin still need the end-to-end ADO run captured on the PR before merge, which stays the human-gated step (the running app is held by another fork and a live board can't be driven here). What changed is that the steering is now explicitly two-sided instead of leaving the symmetric case to the model.

Validation: build ✅, non-DB vitest ✅ (178 passed), biome ✅ on both files.

**Review by calumjs — COMMENTED** [review]
## Review status: blocked — no code review produced

This automated review run could not evaluate PR #901. The Codex review lens failed before producing any findings, so the PR diff itself was not assessed below. Two independent blockers occurred:

### High

**1. Codex review never ran — runtime failed on two fronts**

- **Usage limit hit.** Codex returned: `You've hit your usage limit. Upgrade to Pro ... try again at 8:15 PM.` No review output was generated.
- **Companion crashed with ENOSPC.** Immediately after, the companion crashed while persisting job state:
  `Error: ENOSPC: no space left on device, write` at `Object.writeFileSync (node:fs:2408:20)` → `saveState (.../openai-codex/codex/1.0.2/scripts/lib/state.mjs:114:6)` (Node.js v24.1.0, exit code 1).

  The crash originates at `state.mjs:114`, which is the `fs.writeFileSync(resolveStateFile(cwd), ...)` call inside `saveState`. The state-backing drive is full (C: at 100% use), so the write cannot complete.

**Remedy:**
1. Free disk space on the drive backing the Codex companion state directory (the repo root currently holds multi-GB media files and a `tools-scratch/` directory that account for much of the pressure).
2. Once space is reclaimed, retry after the usage limit resets (reported 8:15 PM) or top up Codex credits at https://chatgpt.com/codex/settings/usage.
3. Re-run the PR #901 review after both blockers are cleared.

Note: this is a tooling/infrastructure blocker, **not** a defect in the PR #901 diff. The change itself was not reviewed and should be re-run once the environment is healthy.

**Review by tomek-i — APPROVED** [review]
